<a href="https://colab.research.google.com/github/mostafa-ja/Anomaly-detection/blob/main/autoencoder5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# we upload, in case that we cant parse logs(a lot of time and ram consumption)

# Mount Google Drive to save datasets
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
#copy log file
!cp '/content/drive/MyDrive/zHDFS2/logs_train' '/content/'
!cp '/content/drive/MyDrive/zHDFS2/logs_ntest' '/content/'
!cp '/content/drive/MyDrive/zHDFS2/logs_atest' '/content/'
!cp '/content/drive/MyDrive/zHDFS2/log2index' '/content/'
!cp '/content/drive/MyDrive/zHDFS2/reduced_embeddings' '/content/'


In [118]:
import pandas as pd
import json
from sklearn.metrics.pairwise import cosine_similarity
import time
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import TensorDataset, DataLoader
import os
import numpy as np


# Device configuration
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

In [4]:
with open('/content/reduced_embeddings', 'r') as json_file:
    embeddings = json.load(json_file)

In [5]:
len(embeddings[0])


10

In [6]:
# all unique templates in our train datasets

dataset = 'logs_train'
templates = set()

with open('/content/' + dataset, 'r') as f:
        for row in f:
          for temp in row.split():
            templates.add(temp)

print(sorted(templates))
print('nember of templates : ',len(templates))

['0', '1', '10', '11', '12', '13', '14', '15', '16', '2', '3', '4', '5', '6', '7', '8', '9']
nember of templates :  17


In [7]:
# all unique templates in our test datasets

datasets = ['logs_ntest','logs_atest']
templates = set()

for dataset in datasets:
  with open('/content/' + dataset, 'r') as f:
          for row in f:
            for temp in row.split():
              templates.add(temp)

print(sorted(templates))
print('nember of templates : ',len(templates))

['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '2', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '3', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '4', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '5', '50', '51', '6', '7', '8', '9']
nember of templates :  52


In [8]:

name = 'logs_train'
window_size = 10
num_sessions = 0
inputs = []
#outputs = []

with open('/content/' + name, 'r') as f:
        for row in f:
            num_sessions += 1
            line = [ embeddings[int(i)] for i in row.strip().split()]
            for i in range(len(line) - window_size):
                inputs.append(line[i:i + window_size])
                #outputs.append(line[i + window_size])

print('Number of sessions({}): {}'.format(name, num_sessions))
print('Number of seqs({}): {}'.format(name, len(inputs)))
#dataset = TensorDataset(torch.tensor(inputs, dtype=torch.float), torch.tensor(outputs))
dataset = TensorDataset(torch.tensor(inputs, dtype=torch.float))

Number of sessions(logs_train): 446578
Number of seqs(logs_train): 4704000


In [9]:
print(inputs[0])

[[-0.3891972303390503, 0.12810246646404266, -0.07853731513023376, -0.10766388475894928, 0.23173417150974274, -0.08228960633277893, 0.017653826624155045, -0.030952107161283493, 0.040218982845544815, 0.030139772221446037], [0.38630372285842896, -0.12673094868659973, -0.11596252024173737, -0.18444840610027313, 0.08815333247184753, -0.02099493145942688, 0.10797911882400513, -0.09289209544658661, -0.05879585072398186, 0.03421926125884056], [-0.3891972303390503, 0.12810246646404266, -0.07853731513023376, -0.10766388475894928, 0.23173417150974274, -0.08228960633277893, 0.017653826624155045, -0.030952107161283493, 0.040218982845544815, 0.030139772221446037], [-0.3891972303390503, 0.12810246646404266, -0.07853731513023376, -0.10766388475894928, 0.23173417150974274, -0.08228960633277893, 0.017653826624155045, -0.030952107161283493, 0.040218982845544815, 0.030139772221446037], [-0.21878856420516968, -0.2136467546224594, 0.04706411808729172, 0.20318950712680817, -0.24529194831848145, 0.02122925035

In [193]:
class Encoder(nn.Module):
    def __init__(self, input_dim, hid_dim, n_layers): #embedding dimensions
        super().__init__()

        self.hid_dim = hid_dim
        self.n_layers = n_layers

        self.rnn = nn.LSTM(input_dim, hid_dim, n_layers, bidirectional=True,batch_first=True)


    def forward(self, src):

        #src = [src len, batch size]


        outputs, (hidden, cell) = self.rnn(src)

        #outputs = [batch size,src len, hid dim * n directions]
        #hidden = [batch size,n layers * n directions,  hid dim]
        #cell = [batch size, n layers * n directions, hid dim]

        #outputs are always from the top hidden layer

        return outputs[:,-1:,:]  #outputs[:,-1:,:]=output of the end of blocks which is concatenated of two directions(we use -1: for keeping the dimensions)

class Decoder(nn.Module):
    def __init__(self, input_dim, hid_dim, n_layers):       #input _dim depends on encoder's output
        super().__init__()

        self.hid_dim = hid_dim
        self.n_layers = n_layers

        self.rnn = nn.LSTM(input_dim, hid_dim, n_layers, bidirectional=False,batch_first=True)

    def forward(self, src):

        #src = [src len, batch size]
        src = torch.tile(src, (1, 10, 1)) # give the output of encoder to each cell of decoder

        outputs, (hidden, cell) = self.rnn(src)

        #outputs = [batch size,src len, hid dim * n directions]
        #hidden = [batch size,n layers * n directions,  hid dim]
        #cell = [batch size, n layers * n directions, hid dim]

        #outputs are always from the top hidden layer

        return outputs

class Autoencoder(nn.Module):
    def __init__(self, encoder, decoder):
        super().__init__()

        self.encoder = encoder
        self.decoder = decoder

    def forward(self, src):
        encoded = self.encoder(src)
        decoded = self.decoder(encoded)
        return decoded

In [194]:
encoder = Encoder(10,32,1)
decoder = Decoder(64,10,1)
model = Autoencoder(encoder, decoder).to(device)
model

Autoencoder(
  (encoder): Encoder(
    (rnn): LSTM(10, 32, batch_first=True, bidirectional=True)
  )
  (decoder): Decoder(
    (rnn): LSTM(64, 10, batch_first=True)
  )
)

In [195]:
input_data = torch.rand(1, 10, 10).to(device) #(batch,sequence_length,embedding)
output= model(input_data)
print(output.shape)

torch.Size([1, 10, 10])


In [196]:
# Calculate the number of parameters
num_params = sum(p.numel() for p in model.parameters())
print(f"Number of Parameters: {num_params}")

Number of Parameters: 14304


In [197]:
dataloader = DataLoader(dataset, batch_size=256)

# Loss and optimizer
criterion = nn.MSELoss()

In [198]:
len(dataloader) # regard to batch size

18375

In [199]:
for step, (seq) in enumerate(dataloader):
  print(seq[0].shape)
  break

torch.Size([256, 10, 10])


In [200]:
def adjust_learning_rate(optimizer, epoch, lr_step=(30, 50), lr_decay_ratio=0.1):
    """Adjust the learning rate based on the epoch number."""
    if epoch == 0:
        optimizer.param_groups[0]['lr'] /= 16
    elif epoch in [1, 2, 3, 4]:  # in step five , we finish warm up ,and start normal learning rate
        optimizer.param_groups[0]['lr'] *= 2
    if epoch in lr_step: # in these steps , we are geting close to optimal point so we need to have shorter step
        optimizer.param_groups[0]['lr'] *= lr_decay_ratio
    return optimizer

optimizer = optim.Adam(model.parameters(), lr=0.001, betas=(0.9, 0.999))

In [201]:

num_epochs = 60
input_size = 10 # embedding size

# Train the model
start_time = time.time()
total_step = len(dataloader)
for epoch in range(num_epochs):  # Loop over the dataset multiple times
    optimizer = adjust_learning_rate(optimizer, epoch)
    train_loss = 0
    for step, (seq) in enumerate(dataloader):
        # Forward pass
        seq = seq[0].clone().detach().view(-1, window_size, input_size).to(device)
        output = model(seq)
        loss = criterion(output, seq.to(device))

        # Backward and optimize
        optimizer.zero_grad()
        loss.backward()
        train_loss += loss.item()
        optimizer.step()
    print('Epoch [{}/{}], train_loss: {:.4f}'.format(epoch + 1, num_epochs, train_loss / total_step))
elapsed_time = time.time() - start_time
print('elapsed_time: {:.3f}s'.format(elapsed_time))
print('Finished Training')


Epoch [1/100], train_loss: 0.0165
Epoch [2/100], train_loss: 0.0116
Epoch [3/100], train_loss: 0.0095
Epoch [4/100], train_loss: 0.0085
Epoch [5/100], train_loss: 0.0067
Epoch [6/100], train_loss: 0.0055
Epoch [7/100], train_loss: 0.0049
Epoch [8/100], train_loss: 0.0046
Epoch [9/100], train_loss: 0.0043
Epoch [10/100], train_loss: 0.0040
Epoch [11/100], train_loss: 0.0038
Epoch [12/100], train_loss: 0.0036
Epoch [13/100], train_loss: 0.0035
Epoch [14/100], train_loss: 0.0034
Epoch [15/100], train_loss: 0.0033
Epoch [16/100], train_loss: 0.0032
Epoch [17/100], train_loss: 0.0032
Epoch [18/100], train_loss: 0.0031
Epoch [19/100], train_loss: 0.0031
Epoch [20/100], train_loss: 0.0030
Epoch [21/100], train_loss: 0.0030
Epoch [22/100], train_loss: 0.0031
Epoch [23/100], train_loss: 0.0030
Epoch [24/100], train_loss: 0.0028
Epoch [25/100], train_loss: 0.0029
Epoch [26/100], train_loss: 0.0028
Epoch [27/100], train_loss: 0.0028
Epoch [28/100], train_loss: 0.0028
Epoch [29/100], train_loss: 0

KeyboardInterrupt: ignored

In [205]:
torch.save(model.state_dict(), '/content/drive/MyDrive/zHDFS2/Autoencoder_parameters.pth')

In [55]:
input_data = seq[0].reshape(1,10,10)
output= model(input_data)
print(output.shape)

torch.Size([1, 10, 10])


In [56]:
a = input_data.cpu().detach().numpy()[0] - output.cpu().detach().numpy()[0]

In [57]:
np.sum(a[5])

0.073634475

In [58]:
torch.sum(seq[0][0]-seq[0][3])

tensor(-0.1362, device='cuda:0')

In [59]:
cosine_similarity(input_data.cpu().detach().numpy()[0],output.cpu().detach().numpy()[0])

array([[ 0.9610373 ,  0.02295969,  0.04059182, -0.00797994,  0.06608871,
        -0.11450423,  0.06367902, -0.538867  , -0.54200083, -0.5351264 ],
       [ 0.06265119,  0.99223757,  0.66928744,  0.9965995 ,  0.65959114,
         0.99361   ,  0.64131004, -0.45647153, -0.4406333 , -0.45387778],
       [ 0.20867422,  0.61992234,  0.99156976,  0.6591899 ,  0.9915739 ,
         0.6456291 ,  0.9953552 , -0.03214208, -0.0064817 , -0.00773428],
       [ 0.06265119,  0.99223757,  0.66928744,  0.9965995 ,  0.65959114,
         0.99361   ,  0.64131004, -0.45647153, -0.4406333 , -0.45387778],
       [ 0.20867422,  0.61992234,  0.99156976,  0.6591899 ,  0.9915739 ,
         0.6456291 ,  0.9953552 , -0.03214208, -0.0064817 , -0.00773428],
       [ 0.06265119,  0.99223757,  0.66928744,  0.9965995 ,  0.65959114,
         0.99361   ,  0.64131004, -0.45647153, -0.4406333 , -0.45387778],
       [ 0.20867422,  0.61992234,  0.99156976,  0.6591899 ,  0.9915739 ,
         0.6456291 ,  0.9953552 , -0.03214208

In [70]:
len(embeddings)

53

In [80]:
embeddings.append([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])

In [ ]:
embeddings.append([-1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0])

In [186]:
embeddings[52]

[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

In [82]:
embeddings[53]

[-1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0]

In [185]:
def generate(name):
    # If you what to replicate the DeepLog paper results(Actually, I have a better result than DeepLog paper results),
    # you should use the 'list' not 'set' to obtain the full dataset, I use 'set' just for test and acceleration.
    hdfs = set()
    # hdfs = []
    with open('/content/' + name, 'r') as f:
        for row in f:
            line = [int(i) for i in row.strip().split()]
            line = line + [52] * (window_size + 1 - len(line)) #if the length of the line is less than windows size, it covers by 30 a template with zeros vector
            hdfs.add(tuple(line))
            # hdfs.append(tuple(line))
    print('Number of sessions({}): {}'.format(name, len(hdfs)))
    return hdfs

In [187]:
test_normal_loader = generate('logs_ntest')
test_abnormal_loader = generate('logs_atest')

Number of sessions(logs_ntest): 1091
Number of sessions(logs_atest): 4126


In [85]:
with torch.no_grad():
    for line in test_normal_loader:
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)
            break
        break

In [89]:
session

(1, 0, 0, 0, 2, 3, 4, 2, 3, 2)

In [112]:
output[:,0,:].shape

torch.Size([1, 10])

In [94]:
np.array(embeddings).shape

(54, 10)

In [111]:
np.argmax(cosine_similarity(output.cpu().detach().numpy()[0,:,:],np.array(embeddings)) ,axis=1 )

array([1, 0, 0, 9, 2, 3, 4, 2, 3, 2])

In [117]:
cosine_similarity(output.cpu().detach().numpy()[:,3,:],np.array(embeddings))

array([[ 0.93364596, -0.4434544 ,  0.21914434,  0.19454281, -0.65147745,
         0.71474941,  0.53797707,  0.15221624, -0.06525082,  0.96791527,
        -0.5763621 , -0.31966809, -0.08581298,  0.3525665 , -0.51343656,
        -0.54429658, -0.57737701,  0.69773972,  0.21558633,  0.38540546,
         0.6757934 ,  0.67971468,  0.53628898,  0.30833949,  0.2453729 ,
         0.46184703,  0.71523791,  0.25231187,  0.29977316,  0.3604768 ,
         0.3592888 ,  0.63598188,  0.18424333,  0.61498045,  0.03822683,
         0.21031395,  0.38700118,  0.67300933,  0.42254237,  0.68351357,
        -0.30130725,  0.29635648, -0.40252998, -0.23298372,  0.7300521 ,
         0.22557606,  0.6841217 ,  0.68081174,  0.69443181, -0.40250046,
         0.18954596,  0.39670442,  0.        ,  0.30450481]])

In [121]:
matrix = cosine_similarity(seq.cpu().detach().numpy()[0],output.cpu().detach().numpy()[0])
np.diag(matrix)

array([0.94757855, 0.99033636, 0.9951027 , 0.93364596, 0.96575195,
       0.9642491 , 0.9754004 , 0.97802216, 0.9820594 , 0.9909105 ],
      dtype=float32)

In [124]:
np.min(np.diag(matrix))

0.9390317

In [127]:
with torch.no_grad():
    c=0
    w=0
    for line in test_normal_loader:
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)
            matrix = cosine_similarity(seq.cpu().detach().numpy()[0],output.cpu().detach().numpy()[0])
            if np.min(np.diag(matrix))<0.9:
              w+=1
            else:
              c+=1

print(w)
print(c)

6396
2334


In [202]:
with torch.no_grad():
    c=0
    w=0
    for line in test_abnormal_loader:
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)
            matrix = cosine_similarity(seq.cpu().detach().numpy()[0],output.cpu().detach().numpy()[0])
            if np.min(np.diag(matrix))<0.9:
              w+=1
            else:
              c+=1

print(w)
print(c)

72244
8122


In [146]:
with torch.no_grad():
    for line in test_normal_loader:
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)
            break
        break

In [173]:
matrix = cosine_similarity(output.cpu().detach().numpy()[0],np.array(embeddings))

In [161]:
matrix[0]

array([-5.85169349e-01,  6.77672409e-01, -3.43758426e-01,  8.08315442e-02,
        9.75183013e-01, -1.96573267e-01, -4.25985141e-01, -2.06848171e-01,
        3.41812066e-02, -6.89816140e-01,  4.21463263e-01, -5.08992339e-02,
       -2.14109347e-01, -5.53116547e-01,  4.62143564e-01,  8.06513570e-02,
        5.66293743e-01, -7.32496992e-01, -4.77402144e-01, -6.20806589e-01,
       -7.31461998e-01, -7.61050253e-01, -5.22182900e-01, -4.42270652e-01,
       -3.76862072e-01, -7.16166436e-01, -7.48330964e-01, -3.50888679e-01,
        4.60564130e-04, -6.21488362e-01, -4.53857385e-01, -6.49230320e-01,
       -4.85013152e-01, -7.63704227e-01, -4.39897555e-01, -4.54773786e-01,
       -5.11414433e-01, -6.74415146e-01, -5.51735490e-01, -7.30483624e-01,
       -5.72971238e-02, -4.74739104e-01,  1.27643124e-01, -2.77818346e-01,
       -7.62999968e-01, -3.62153647e-01, -6.76494860e-01, -7.93673361e-01,
       -7.61930970e-01,  3.70575304e-02, -4.91890402e-01, -7.18702580e-01,
        0.00000000e+00, -

In [170]:
session

(4, 4, 11, 10, 10, 10, 10, 11, 11, 15)

In [174]:
a = np.argsort(matrix)
a

array([[47, 33, 44, 48, 21, 26, 17, 20, 39, 51, 25,  9, 46, 37, 31, 29,
        19,  0, 13, 38, 22, 36, 50, 32, 18, 41, 35, 30, 23, 34,  6, 24,
        45, 27,  2, 43, 12,  7,  5, 40, 11, 53, 52, 28,  8, 49, 15,  3,
        42, 10, 14, 16,  1,  4],
       [26, 33, 21, 48,  0,  9, 44, 47, 20, 17, 39, 13, 37, 46, 31,  5,
        25, 34, 51, 29,  6, 19, 12, 11, 38, 40, 36, 49, 43, 28, 22, 15,
        52, 30,  7, 53,  8, 32, 35, 50, 18, 42, 23, 41, 24, 45, 14, 27,
         2, 10, 16,  3,  1,  4],
       [ 9, 21,  0, 26, 48, 44, 47, 33, 20, 17, 39, 37, 46, 31, 13,  5,
         6, 25, 51, 29, 19, 36, 38, 34, 22, 30, 28, 53,  7, 23, 35, 41,
        50, 32, 24, 18, 45, 43, 12, 27, 52,  2, 49,  8,  3, 40, 11, 15,
        42, 14, 16, 10,  1,  4],
       [ 9, 44, 21,  0, 17, 47, 48, 39, 26, 46, 37, 20, 25, 33, 51, 31,
        22, 19, 38,  6, 29, 36,  5, 13, 30, 23, 41, 50, 18, 32, 27, 35,
        24, 45,  2, 34,  7, 28,  3, 53, 43, 12, 52,  8, 49, 11, 40, 15,
        14, 42, 16, 10,  1,  4],
    

In [175]:
a[1]

array([26, 33, 21, 48,  0,  9, 44, 47, 20, 17, 39, 13, 37, 46, 31,  5, 25,
       34, 51, 29,  6, 19, 12, 11, 38, 40, 36, 49, 43, 28, 22, 15, 52, 30,
        7, 53,  8, 32, 35, 50, 18, 42, 23, 41, 24, 45, 14, 27,  2, 10, 16,
        3,  1,  4])

In [176]:
np.where(a[1]==session[1])[0][0]

53

In [177]:
matrix = cosine_similarity(output.cpu().detach().numpy()[0],np.array(embeddings))
a = np.argsort(matrix)
s=[]
for i in range(10):
  s.append(np.where(a[i]==session[i])[0][0])

In [178]:
s

[53, 53, 46, 51, 53, 53, 52, 52, 47, 41]

In [203]:
with torch.no_grad():
    for line in test_normal_loader:
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            matrix = cosine_similarity(output.cpu().detach().numpy()[0],np.array(embeddings))
            a = np.argsort(matrix)
            s=[]
            for i in range(10):
              s.append(np.where(a[i]==session[i])[0][0])
            print(s)
        print('==========') #to show this block finished

[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 52, 53, 53, 53, 53, 53, 11]
[53, 53, 52, 53, 53, 53, 53, 53, 8, 19]
[53, 53, 53, 53, 53, 53, 53, 29, 3, 13]
[52, 53, 53, 53, 53, 53, 5, 15, 40, 19]
[53, 53, 53, 53, 53, 34, 14, 45, 18, 34]
[52, 53, 53, 53, 9, 18, 45, 3, 12, 50]
[52, 53, 53, 38, 16, 46, 19, 31, 50, 46]
[52, 53, 53, 53, 50, 50, 53, 52, 48, 52]
[53, 53, 53, 50, 52, 51, 52, 53, 51, 53]
[53, 52, 51, 51, 53, 51, 53, 52, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 48, 53, 45]
[53, 53, 53, 53, 53, 53, 53, 53, 48, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[52, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 52, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 

KeyboardInterrupt: ignored

In [204]:
with torch.no_grad():
    for line in test_abnormal_loader:
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            matrix = cosine_similarity(output.cpu().detach().numpy()[0],np.array(embeddings))
            a = np.argsort(matrix)
            s=[]
            for i in range(10):
              s.append(np.where(a[i]==session[i])[0][0])
            print(s)
        print('==========') #to show this block finished

[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 52, 53, 53, 53, 53, 53, 53]
[53, 53, 52, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 51, 53, 53, 41]
[53, 53, 53, 53, 53, 53, 53, 53, 42, 26]
[53, 53, 53, 53, 53, 53, 53, 30, 17, 21]
[53, 53, 53, 53, 53, 53, 44, 13, 39, 1]
[52, 53, 53, 53, 53, 7, 25, 6, 9, 53]
[53, 53, 53, 53, 39, 16, 40, 47, 46, 21]
[53, 53, 53, 38, 19, 32, 51, 0, 5, 39]
[53, 53, 28, 23, 3, 53, 51, 15, 12, 49]
[53, 39, 14, 26, 0, 1, 1, 36, 21, 40]
[40, 15, 23, 24, 53, 16, 25, 6, 43, 52]
[15, 39, 53, 53, 5, 39, 5, 53, 40, 44]
[47, 53, 53, 18, 38, 5, 53, 43, 53, 17]
[53, 53, 19, 27, 4, 53, 47, 47, 19, 15]
[53, 16, 29, 4, 53, 50, 41, 15, 38, 53]
[22, 48, 5, 49, 38, 35, 22, 15, 3, 4]
[39, 6, 53, 42, 52, 20, 51, 53, 53, 53]
[35, 53, 41, 51, 53, 42, 50, 45, 48, 9]
[53, 40, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 53]
[53, 53, 53, 53, 53, 53, 53, 53, 53, 

KeyboardInterrupt: ignored

In [213]:
s = [53, 50, 53, 53, 43, 51, 53, 42, 53, 23]
np.sum(53 - np.array(s) > 6)

3

In [219]:
with torch.no_grad():
    for line in test_normal_loader:
        score = 0
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            matrix = cosine_similarity(output.cpu().detach().numpy()[0],np.array(embeddings))
            a = np.argsort(matrix)
            s=[]
            for i in range(10):
              s.append(np.where(a[i]==session[i])[0][0])

            score += np.sum(53 - np.array(s) > 20)
        print(score/(len(line) - window_size))


1.6363636363636365
0.0
0.0
0.0
0.0
2.0
1.6363636363636365
0.0
0.0
1.0909090909090908
0.0
0.1111111111111111
0.0
0.0
1.3636363636363635
0.6666666666666666
0.1111111111111111
0.8888888888888888
0.3333333333333333
0.0
0.0
0.5555555555555556
0.0
0.0
1.6666666666666667
0.2222222222222222
0.0
0.0
0.0
0.0
0.0
1.6666666666666667
0.0
0.0
0.0
0.0
0.0
0.8888888888888888
0.2222222222222222
0.0
0.1111111111111111
1.9166666666666667
0.0
0.5555555555555556
1.2222222222222223
0.3333333333333333
0.1111111111111111
1.5555555555555556
1.6666666666666667
0.2222222222222222
0.0
0.0
0.3333333333333333
0.0
0.3333333333333333
0.0
0.0
0.0
0.3333333333333333
0.2222222222222222
0.0
0.0
0.0
0.0
0.0
0.1111111111111111
0.0
0.2222222222222222
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.8181818181818181
0.6666666666666666
0.2
0.0
0.0
1.5454545454545454
0.6666666666666666
1.8
0.0
0.0
1.4444444444444444
0.3333333333333333
0.3333333333333333
0.0
0.0
0.0
0.2222222222222222
0.25
0.0
0.5
0.0
0.0
0.0
0.0
0.1111111111111111
0.4444

In [220]:
with torch.no_grad():
    for line in test_abnormal_loader:
        score = 0
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            matrix = cosine_similarity(output.cpu().detach().numpy()[0],np.array(embeddings))
            a = np.argsort(matrix)
            s=[]
            for i in range(10):
              s.append(np.where(a[i]==session[i])[0][0])

            score += np.sum(53 - np.array(s) > 20)
        print(score/(len(line) - window_size))

2.7142857142857144
0.38095238095238093
1.6
1.6666666666666667
0.75
0.0
2.6666666666666665
2.392857142857143
0.0
2.1666666666666665
2.5483870967741935
2.1842105263157894
0.8421052631578947
1.9473684210526316
2.4074074074074074
2.8333333333333335
2.466666666666667
2.111111111111111
3.0588235294117645
0.3076923076923077
0.21428571428571427
3.0588235294117645
2.411764705882353
0.3333333333333333
2.65625
2.2777777777777777
2.392857142857143
3.2857142857142856
0.30434782608695654
0.1
2.6333333333333333
2.0588235294117645
1.1538461538461537
0.0
0.5294117647058824
0.1
1.631578947368421
1.0
1.4444444444444444
2.696969696969697
0.0
2.0
2.25
0.47368421052631576
0.42857142857142855
0.35
2.5357142857142856
3.1052631578947367
2.59375
2.7096774193548385
2.71875
2.8947368421052633
2.375
2.5
1.9411764705882353
1.1764705882352942
1.3571428571428572
0.7894736842105263
0.06666666666666667
0.23076923076923078
0.9285714285714286
0.7272727272727273
0.3333333333333333
2.0
1.3846153846153846
1.4736842105263157

KeyboardInterrupt: ignored

In [272]:
with torch.no_grad():
    c=0
    for line in test_normal_loader:
        sum_loss = 0
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            loss = criterion(output, seq)

            sum_loss += loss.cpu().detach().numpy()

        score = (sum_loss/(len(line) - window_size))
        if (score>0.01):
          c+=1

print(c)

241


In [274]:
with torch.no_grad():
    c=0
    for line in test_abnormal_loader:
        sum_loss = 0
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            loss = criterion(output, seq)

            sum_loss += loss.cpu().detach().numpy()

        score = (sum_loss/(len(line) - window_size))
        if (score<0.01):
          c+=1

print(c)


1189



Number of sessions(logs_ntest): 1091

Number of sessions(logs_atest): 4126

In [268]:
with torch.no_grad():
    c=0
    for line in test_normal_loader:
        score = 0
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            loss = criterion(output, seq)

            if score < loss.cpu().detach().numpy():
              score = loss.cpu().detach().numpy()
        if (score>0.01):
          c+=1

print(c)

511


In [266]:
with torch.no_grad():
    c=0
    for line in test_abnormal_loader:
        score = 0
        for i in range(len(line) - window_size):
            session = line[i:i + window_size]
            seq = [embeddings[temp] for temp in session]
            seq = torch.tensor(seq, dtype=torch.float).view(-1, window_size, input_size).to(device)
            output = model(seq)

            loss = criterion(output, seq)

            if score < loss.cpu().detach().numpy():
              score = loss.cpu().detach().numpy()
        if (score<0.01):
          c+=1

print(c)

152
